Python/ceval.c

```diff
@@ -3560,33 +3598,92 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, InterpreterFrame *frame, int thr
             DISPATCH();
         }
 
-        TARGET(JUMP_IF_NOT_EXC_MATCH) {
-            const char *cannot_catch_msg = "catching classes that do not "
-                                           "inherit from BaseException is not "
-                                           "allowed";
-            PyObject *right = POP();
-            PyObject *left = TOP();
-            if (PyTuple_Check(right)) {
-                Py_ssize_t i, length;
-                length = PyTuple_GET_SIZE(right);
-                for (i = 0; i < length; i++) {
-                    PyObject *exc = PyTuple_GET_ITEM(right, i);
-                    if (!PyExceptionClass_Check(exc)) {
-                        _PyErr_SetString(tstate, PyExc_TypeError,
-                                         cannot_catch_msg);
-                        Py_DECREF(right);
-                        goto error;
-                    }
-                }
+        TARGET(JUMP_IF_NOT_EG_MATCH) {
+            PyObject *match_type = POP();
+            PyObject *exc_type = TOP();
+            PyObject *exc_value = SECOND();
+            if (!check_except_star_type_valid(tstate, match_type)) {
+                Py_DECREF(match_type);
+                goto error;
+            }
+
+            PyObject *match = NULL, *rest = NULL;
+            int res = exception_group_match(
+                exc_type, exc_value, match_type, &match, &rest);
+            Py_DECREF(match_type);
+            if (res == -1) {
+                goto error;
+            }
+
+            if (match == NULL || rest == NULL) {
+                Py_XDECREF(match);
+                Py_XDECREF(rest);
```

Comment:
AFAICT, if one of these is NULL, so is the other.

```suggestion
                assert(match == NULL);
                assert(rest == NULL);
                Py_DECREF(match);
                Py_DECREF(rest);
```